RocketChat/Rocket.Chat
# [FIX][ENTERPRISE] Engagement Dashboard displaying incorrect data about active users

## Proposed changes (including videos or screenshots)


 - Fix sessions' and users' grouping in the Engagement Dashboard API endpoints;
 - Fix the data displayed in the charts from the "Active users", "Users by time of day" and "When is the chat busier?" sections of the Engagement Dashboard;
 - Replace label used to describe the amount of Active Users in the License section of the Info page.



## Issue(s)

[Task - ClickUp](https://app.clickup.com/t/5131xn)

## Steps to test or reproduce

- Running an Enterprise Version's instance of Rocket.Chat and logged in as an admin, open Administration > **Info**. The license card should display the amount of **Active** users (which may be limited to a given number) -- active users do not include app users, guests and deactived users.
- Running an Enterprise Version's instance of Rocket.Chat and logged in as an admin, open Administration > **Engagement Dashboard**. All the data shown in the "Active users", "Users by time of day" and "When is the chat busier?" sections must be limited to the amount of active users displayed in the Info page.

## Further comments

### **Amount of users in the Info page**
the `Users` label has been replaced by `Active_Users` since only the amount of **active** users is limited in the EE (the pie chart was already comparing the amount of **active** users with the maximum amount of active users, just the label's change was needed for clarity);

### **"Active users" section in the Engagement Dashboard**
 - **Previous behavior**: the API endpoint returned an object containing dates and the amount of sessions opened in each date (`users` field). Since a user can start multiple sessions in a same date, the value of the `users` field could exceed the amount of active users in the server (_incorret behavior_).
 - **New behavior**: the API endpopint now returns an object containg dates and, for each date, a list of users (`usersList`, which is a **set** of userIds) and the `users` field (size of the `usersList` field). The API endpoint now groups all sessions by date **and `userId`** and then perform one more group to sum the amount of distinct active users in a date.
_**Front-end:**_ The amount of weekly and monthly active users used to be calculated summing the amount of daily active users in an interval. However, recurring DAU could be counted multiple times (exceeding the amount of active users in the server). The amount of MAU and WAU obtained with the help of a **set** that aggregates all distinct `userId`s (related to active sessions) in a period -- in order to assure that only **distinct** users are being counted.

### **"Users by time of day" section in the Engagement Dashboard**
 - **Previous behavior**: the API endpoint returned a `users` field associated with each 1 hour interval between the beginning and end date. The `users` field was the amount of active sessions in each 1 hour interval. However, since a user can start multiple sessions in a same hour, the value of the `users` field could exceed the amount of active users in the server (_incorret behavior_).
 - **New behavior**: the API endpopint now returns list of users (`usersList`, which is a **set** of userIds) and the `users` field (size of the `usersList` field) associated with each 1 hour interval. The API endpoint now groups all distinct `userId`s that created sessions in a same range (1 hour interval) and then add the `users` field (which is the size of each range set).

### **"When is the chat busier?" (Hours) section in the Engagement Dashboard**
 - **Previous behavior**: since each bar in the chart covers a 2 hour range, the sum of two adjacent ranges was used as the `y` value of each bar. However, since a user can have one same session active for more than one hour, these cases were being counted twice. Also, if a user opened multiple sessions in a 2 hours range, these sessions wouldn't be grouped and they would be counted as if they were from distinct users.
 - **New behavior**: applies the same new approach used in the "Users by time of day" section, but using intervals (ranges) of 2 hours.
